### PR TITLE
Upgrading Jest to v29 and optimizing default configuration

### PR DESCRIPTION
**Pull request by ymqy**
## Summary

I am preparing to split into smaller separate pr's to make review and merge easier, look https://github.com/facebook/react/pull/26088

## How did you test this change?
ci green

### CODE REVIEW

**Comment by eps1lon** [comment]
There is also https://github.com/facebook/react/pull/21575 if you need help with certain CI checks

**Comment by ymqy** [comment]
> There is also #21575 if you need help with certain CI checks

Thank you for bringing that to my attention, I will check out #21575 and resolve these CI errors.

File: scripts/jest/spec-equivalence-reporter/equivalenceReporter.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ */
+
+'use strict';
+
+class EquivalenceReporter {
+  onTestCaseResult(test, testCaseResult) {
+    console.log(
+      `EQUIVALENCE: ${testCaseResult.title}, ` +
+        `status: ${testCaseResult.status}, ` +
+        `numExpectations: ${testCaseResult.numPassingAsserts}`
```
Comment (ymqy) [inline scripts/jest/spec-equivalence-reporter/equivalenceReporter.js:15]: Waiting for release of https://github.com/facebook/jest/pull/13795

**Comment by ymqy** [comment]
Tried to run my own project branch (ymqy:feat/upgrade-jest) in Circle CI, not sure why it affects the current PR CI check

**Comment by ymqy** [comment]
Currently there are two problems blocking
- restore node_modules cache step does not find the cache resulting in "Cannot find module 'chalk'" error, it is not clear what changes are related (https://app.circleci.com/pipelines/github/facebook/react/37549/workflows/ea533b33-4d84-4f8a-81cb-446fc8f6ff35/jobs/617198)
- JavaScript heap out of memory, still in the troubleshooting (https://app.circleci.com/pipelines/github/facebook/react/37549/workflows/ea533b33-4d84-4f8a-81cb-446fc8f6ff35/jobs/617201/parallel-runs/5)

File: scripts/jest/spec-equivalence-reporter/equivalenceReporter.js (hunk 1 above)
Comment (ymqy) [reply scripts/jest/spec-equivalence-reporter/equivalenceReporter.js:15]: [[v29.4.1](https://github.com/facebook/jest/releases/tag/v29.4.1)](https://github.com/facebook/jest/releases/tag/v29.4.1)

**Comment by ymqy** [comment]
> ReactSuspenseWithNoopRenderer-test.js caused OOM when running this test file, and after analysis, I found that these two tests are the cause of the OOM problem, More specific reasons are under further debugging analysis
> 
> * "keeps showing an avoided parent fallback if it is already showing"
> * "do not show placeholder when mounting an avoided boundary with startTransition"

The most fundamental reason is that after the `expect.toEqual` method fails to execute the comparison
https://github.com/facebook/react/blob/0652bdbd10f6192e10f6cab9602f6bbfe5fd2c8e/packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js#L2431

jest needs to print the diff information，it will perform a recursive traversal
https://github.com/facebook/jest/blob/bc84c8a15649aaaefdd624dc83824518c17467ed/packages/expect/src/matchers.ts#L638

and after executing deepCyclicCopyReplaceable(replacedReceived)
https://github.com/facebook/jest/blob/bc84c8a15649aaaefdd624dc83824518c17467ed/packages/jest-matcher-utils/src/index.ts#L410-L422

it enters the recursive loop, because the array element passed into expect contains the fiber property

image

The WeakMap in deepCyclicCopyReplaceable takes up a lot of memory until it triggers OOM. 

image

The main branch code does not have this problem because the testRunner uses jasmine, an exception is thrown after executing deepCyclicCopyReplaceable(replacedReceived), the test case itself is expecting a failed result and just happens to get the passed result (but is this the expected behavior? @acdlite ). After analyzing the call stack, I found that when the testRunner is jasmine, the recursive path is some fiber node -> baseState -> cache -> controller -> Symbol(impl) -> signal -> _globalObject -> jasmine -> process - > mainModule -> some module, then it generates an exception error

image

After migrating from jasmine to circus there is no more jasmine property, it no longer generates an exception error but enters a long recursive loop until OOM.

File: packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js [hunk 2]
```diff
@@ -2428,17 +2440,21 @@ describe('ReactSuspenseWithNoopRenderer', () => {
       'Loading B...',
     ]);
     // Still suspended.
-    expect(ReactNoop.getChildren()).toEqual([span('A')]);
+    expect(removeNonEnumerableProperties(ReactNoop.getChildren())).toEqual([
```
Comment (ymqy) [inline packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js:2443]: Optimization reduced test time from a long time to 9 seconds.

**Comment by ymqy** [comment]
> There is also #21575 if you need help with certain CI checks

@eps1lon 🟢 🥳

**Comment by ymqy** [comment]
@gaearon This PR is important because Jest v29 has been released and the new version contains a lot of new features and bug fix which is important for our project. It's ready for review and merge.

**Comment by gaearon** [comment]
Looks like CI is failing

**Comment by ymqy** [comment]
> Looks like CI is failing

Because No cache is found, occasionally encountered, but it is not clear why

image

**Comment by ymqy** [comment]
> Looks like CI is failing

I reset the commit and then repushed it in order to trigger the ci rerun, and now the test cases all pass again

**Pull request closed without merging**